z3c/dependencychecker/db.py

```diff
@@ -43,3 +48,186 @@ def _filter_duplicates(self, imports):
+            if not self._filter_requirements_lists(testing_imports, x)
+        ]
+        unique = self._get_unique_imports(
+            imports_list=should_be_test_requirements,
+        )
+        return unique
+
+    def get_unneeded_test_requirements(self):
+        test_requirements = self._get_test_extra()
+        if not test_requirements:
+            return []
+
+        filters = (
+            self._filter_known_packages,
+            self._filter_python_standard_library,
+            self._filter_used_imports,
+        )
+        unneeded = self._process_pipeline(test_requirements, filters)
+        unique_imports = self._get_unique_imports(imports_list=unneeded)
+        return unique_imports
+
+    def _process_pipeline(self, objects, filters):
+        final_list = []
+        for single_object in objects:
+            for filter_func in filters:
+                if not filter_func(single_object):
+                    break
+            else:
+                final_list.append(single_object)
+
+        return final_list
+
+    def _get_unique_imports(self, imports_list=None):
+        if imports_list is None:
+            imports_list = self.imports_used
+        unique_dotted_names = list(set(imports_list))
+        sorted_unique_dotted_names = sorted(unique_dotted_names)
+        return sorted_unique_dotted_names
+
+    def _filter_used_imports(self, dotted_name):
+        return self._filter_requirements_lists(self.imports_used, dotted_name)
+
+    def _filter_known_packages(self, dotted_name):
+        to_ignore = (
+            DottedName('setuptools'),
+            DottedName('pkg_resources'),
+            DottedName('distribute'),
+        )
+        return self._filter_requirements_lists(to_ignore, dotted_name)
+
+    @staticmethod
+    def _filter_testing_imports(dotted_name):
+        return not dotted_name.is_test
+
+    @staticmethod
+    def _filter_only_testing_imports(dotted_name):
+        return dotted_name.is_test
+
+    def _filter_own_package(self, dotted_name):
+        own_package = self._build_own_package_import()
+        return dotted_name not in own_package
+
+    def _build_own_package_import(self):
+        return DottedName(self.package_name)
+
+    def _filter_requirements(self, dotted_name):
+        return self._filter_requirements_lists(
+            self._requirements,
+            dotted_name,
+        )
+
+    def _filter_test_requirements(self, dotted_name):
+        test_requirements = self._get_test_extra()
+        if not test_requirements:
+            return True
+
+        return self._filter_requirements_lists(
+            test_requirements,
+            dotted_name,
+        )
+
+    def _filter_python_standard_library(self, dotted_names):
+        std_library = self._build_std_library()
+        return self._filter_requirements_lists(std_library, dotted_names)
+
+    @staticmethod
+    def _filter_requirements_lists(requirements, candidate):
```

Comment:
Make it clearer what happens here. 

- I see the word "candidate" for the first time here.

- Assuming "r" is a single "requirement", what is the meaning of "candidate in requirement"? I'd expect "candidate == requirement"? But that's weird, too.

- "return not any" is not something I can figure out right away. Something with double negatives again, I guess :-) Using an intermediary variable name should make it clearer.

Reply:
I had to battle with this one for quite a bit, let's see if the changes I made make it better:
- I renamed it to ``_discard_if_found_obj_in_list`` and gets two parameters, the object (a ``DottedName``) and a list (of ``DottedName``s)
- I renamed ``candidate`` to ``obj``

Regarding the logic.

I'm using a few tricks of generators:
- ``any`` consumes an iterator and returns True as soon as one element on the iterable is True (see [the official docs](https://docs.python.org/3/library/functions.html#any))
- with that in mind, we pass a generator to ``any`` that tests if the object is in the requirement we are iterating over, as it is a generator, we don't need to go through all possible 300k objects in the list to return, but as soon as one returns True, we are done.
- ``not any`` is because we want to return ``True`` when the object *is found*
- I added plenty of comments and a docstring

Reply:
I haven't used `any` yet. Sounds quite useful. I must admit to not using generators as much as I probably should :-)

But: "any returns True as soon as one element is True" plus "`not any` because we want to return True if found" => ??? I'll have to look at the implementation again, I guess, because this sounds like you return False if anything is found.